### ISSUE TITLE
How to register decorator classes?

### ISSUE THREAD

**Issue by davidkeaveny** [ipjohnson/Grace#149]
If I have an abstract base class `CommandHandler<T>`, and I want to register a decorator class `LoggingCommandHandler<T>` such that every class that implements the base class is resolved with the decorator class, how do I tell Grace to achieve such a thing?

How about if I want to add another decorator class `ValidatingCommandHandler<T>` and I want the order of the decorator classes to be preserved? i.e. always resolve as:

```
LoggingCommandHandler
--> ValidatingCommandHandler
----> ConcreteCommandHandler
```

**Comment by ipjohnson** [comment]
Hi @davidkeaveny 

I took a quick look and out of the box you can't do what you want to do but I think it's possible to do it with a C# extension.

I'll write up an example of how to do it and post a gist of it.

I also think this is a worth feature request so I'll see about making it part of Grace officially so you don't have to do an extension.

**Comment by ipjohnson** [comment]
@davidkeaveny did the unit test for dectorators work for you? Also what are your thoughts on C# extension vs a member on the interface?

**Comment by davidkeaveny** [comment]
It works flawlessly 👍 

Would the extension method basically replace the `.When.MeetsCondition(new WhenDecoratingCondition(typeof(BaseCommand<>)))`? It would certainly clean up the registration process a little more, but it's not essential.

**Comment by davidkeaveny** [comment]
Hmm, maybe I jumped the gun a little bit - the first decorator that I add works flawlessly (the `LoggingCommandHandlerDecorator<>` in my example), but when I then add the registration for another decorator (the `ValidatingCommandHandlerDecorator<>`) then Grace fails to resolve the type. 

I've tried switching it round, and get the same result - register one decorator, all is good; register two decorators, all is bad.

**Comment by ipjohnson** [comment]
Hmm, I can see the piece of code that must have failed to end up in that situation (it's having an issue building the closed generic type for activation).

Would it be possible for you to create a small sample using your types that you want to decorate and the decorators that are failing? You can remove the logic from the classes as I think it's probably some corner case of generics that I'm missing but I'm not able to create it using my simple test cases.

**Comment by ipjohnson** [comment]
Thanks @davidkeaveny  that was very helpful. I was able to replicate the problem and fix it. I should be able to find some time this weekend to look at the other generic issue you raised but I'm not sure what/if there is a fix for it.

Either way I should be able to get a new version out for you this weekend. In the mean time if you want you can try the nightly build at https://ci.appveyor.com/nuget/grace-master

**Comment by ipjohnson** [comment]
I've released version 6.3.3 to address this issue and I'm marking it closed

**Comment by davidkeaveny** [comment]
Looks great, @ipjohnson , thanks!
